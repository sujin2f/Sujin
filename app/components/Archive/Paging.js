import React, { Component } from 'react';
import { withRouter } from 'react-router';

import Link from 'src/components/Link';
import PAGING from 'app/constants/paging';

class Paging extends Component {
  render() {
    const {
      totalPage,
      match: { params },
    } = this.props;

    const entities = [];
    const currentPage = parseInt(params.paged, 10) || 1;
    let url = '';

    if (totalPage !== 0) {
      const start = (currentPage - PAGING.offset) < 1 ? 1 : currentPage - PAGING.offset;
      const end = (currentPage + PAGING.offset) > totalPage
        ? totalPage :
        currentPage + PAGING.offset;

      if (start > PAGING.offset) {
        entities.push(1);
        entities.push('...');
      }

      for (let index = start; index < start + (PAGING.offset * 2) + 1 && index <= end; index += 1) {
        entities.push(index);
      }

      if (end < totalPage - PAGING.offset) {
        entities.push('...');
        entities.push(totalPage);
      }
    }

    if (params.category) {
      url = `category/${params.category}`;
    } else if (params.tag) {
      url = `tag/${params.tag}`;
    } else if (params.searchString) {
      url = `search/${params.searchString}`;
    }

    return (
      <section className="paging row">
        <div className="columns small-12">
          <ul className="pagination text-center" role="navigation" aria-label="Pagination">
            {entities.map(entity => (
              <li className={`${currentPage === entity ? 'active' : ''}`} key={`paging-${entity}`}>
                {entity !== '...' &&
                  <Link href={`${process.env.SUJIN_BASE_URL}${url}/page/${entity}`}>
                    {entity}
                  </Link>
                }
                {entity === '...' &&
                  <span className="fa fa-ellipsis-h" aria-hidden="true" />
                }
              </li>
            ))}
          </ul>
        </div>
      </section>
    );
  }
}

export default withRouter(Paging);
