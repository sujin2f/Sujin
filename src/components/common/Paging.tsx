import React, { Fragment } from 'react'

import { Link } from 'components/common/Link'
import { PAGE_OFFSET } from 'constants/common'

interface Props {
  totalPages: number
  urlPrefix: string
  currentPage: number
}

export const Paging = (props: Props): JSX.Element => {
  const {
    totalPages,
    urlPrefix,
    currentPage,
  } = props

  let entities: number[] = []

  if (totalPages !== 1) {
    const start = (currentPage - PAGE_OFFSET) > 2 ? currentPage - PAGE_OFFSET : 1
    const end = (currentPage + PAGE_OFFSET) < (totalPages - 1) ?
      currentPage + PAGE_OFFSET :
      totalPages

    if (start > 2) {
      entities.push(1)
      entities.push(-1)
    }

    entities = [
      ...entities,
      ...Array.from(Array(end - start + 1).keys()).map((v) => v + start),
    ]

    if (end < totalPages - 1) {
      entities.push(-1)
      entities.push(totalPages)
    }
  }

  return (
    <nav className="columns small-12 paging" role="navigation" aria-label="Paging">
      {entities.map((entity) => {
        const url = `${urlPrefix}/page/${entity}`
        const isCurrnet = currentPage === entity
        return (
          <Fragment key={`paging-${entity}`}>
            {isCurrnet && (
              <span className="active">{entity}</span>
            )}

            {entity !== -1 && !isCurrnet && (
              <Link to={url}>
                {entity.toString()}
              </Link>
            )}

            {entity === -1 && (
              <div className="hellip">
                <span />
                <span />
                <span />
              </div>
            )}
          </Fragment>
        )
      })}
    </nav>
  )
}
