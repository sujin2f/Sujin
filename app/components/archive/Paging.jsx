import hash from 'object-hash';

import Link from 'app/components/router/Link';

import { getPaging } from 'app/utils/archive';
import { isMobile } from 'app/utils/common';

const { Component } = wp.element;

class Paging extends Component {
  render() {
    const {
      totalPages,
      currentPage,
      urlPrefix,
    } = this.props;

    const pagingRange = isMobile() ? 1 : 5;
    const entities = getPaging(totalPages, currentPage, pagingRange);

    return (
      <section className="paging row">
        <div className="columns small-12">
          <ul className="pagination text-center" role="navigation" aria-label="Pagination">
            {entities.map((entity, id) => {
              const url = `${urlPrefix}/page/${entity}`;
              const isCurrnet = currentPage === entity;
              return (
                <li
                  className={`${isCurrnet ? 'active' : ''}`}
                  key={hash(`url-${id}`)}
                >
                  {isCurrnet && (
                    <span>{entity}</span>
                  )}

                  {entity !== -1 && !isCurrnet && (
                    <Link to={url}>
                      {entity}
                    </Link>
                  )}

                  {entity === -1 && (
                    <span>&hellip;</span>
                  )}
                </li>
              );
            })}
          </ul>
        </div>
      </section>
    );
  }
}

export default Paging;
