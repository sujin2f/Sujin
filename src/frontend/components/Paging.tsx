import React, { Fragment } from 'react'

import { Link } from 'src/frontend/components'
import { PAGE_OFFSET } from 'src/frontend/constants/common'

interface Props {
    totalPages: number
    urlPrefix: string
    currentPage: number
}

export const Paging = (props: Props): JSX.Element => {
    const { totalPages, urlPrefix, currentPage } = props

    let entities: number[] = []

    if (totalPages !== 1) {
        const start =
            currentPage - PAGE_OFFSET > 2 ? currentPage - PAGE_OFFSET : 1
        const end =
            currentPage + PAGE_OFFSET < totalPages - 1
                ? currentPage + PAGE_OFFSET
                : totalPages

        if (start > 2) {
            entities.push(1)
            entities.push(-1)
        }

        entities = [
            ...entities,
            ...Array.from(Array(end - start + 1).keys()).map((v) => v + start),
        ]

        if (end < totalPages - 1) {
            entities.push(-1)
            entities.push(totalPages)
        }
    }

    return (
        <nav
            className="paging__container"
            role="navigation"
            aria-label="Navigate to another page"
        >
            {entities.map((entity) => {
                const url = `${urlPrefix}/page/${entity}`
                const isCurrnet = currentPage === entity
                return (
                    <Fragment key={`paging-${entity}`}>
                        {isCurrnet && (
                            <span className="paging paging--active">
                                {entity}
                            </span>
                        )}

                        {entity !== -1 && !isCurrnet && (
                            <Link to={url} className="paging">
                                {entity.toString()}
                            </Link>
                        )}

                        {entity === -1 && (
                            <div className="paging paging--hellip">
                                <span />
                                <span />
                                <span />
                            </div>
                        )}
                    </Fragment>
                )
            })}
        </nav>
    )
}
