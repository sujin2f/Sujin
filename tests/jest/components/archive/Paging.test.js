import React from 'react';
import Paging from '../../../app/components/archive/Paging';
import renderer from 'react-test-renderer';

test('Link changes the class when hovered', () => {
  let component = renderer.create(
    <Paging totalPages="100" currentPage="3" urlPrefix="category/test" />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  component = renderer.create(
    <Paging totalPages="10" currentPage="5" urlPrefix="category/test" />,
  );
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
