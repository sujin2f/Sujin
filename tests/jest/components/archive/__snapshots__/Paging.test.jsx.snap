// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paging 1`] = `
<section
  className="paging row"
>
  <div
    className="columns small-12"
  >
    <ul
      aria-label="Pagination"
      className="pagination text-center"
      role="navigation"
    >
      <li
        className=""
      >
        <a
          href="category/test/page/1"
          onClick={[Function]}
        >
          1
        </a>
      </li>
      <li
        className=""
      >
        <span
          aria-hidden="true"
          className="fa fa-ellipsis-h"
        />
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/45"
          onClick={[Function]}
        >
          45
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/46"
          onClick={[Function]}
        >
          46
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/47"
          onClick={[Function]}
        >
          47
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/48"
          onClick={[Function]}
        >
          48
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/49"
          onClick={[Function]}
        >
          49
        </a>
      </li>
      <li
        className="active"
      >
        <a
          href="category/test/page/50"
          onClick={[Function]}
        >
          50
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/51"
          onClick={[Function]}
        >
          51
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/52"
          onClick={[Function]}
        >
          52
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/53"
          onClick={[Function]}
        >
          53
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/54"
          onClick={[Function]}
        >
          54
        </a>
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/55"
          onClick={[Function]}
        >
          55
        </a>
      </li>
      <li
        className=""
      >
        <span
          aria-hidden="true"
          className="fa fa-ellipsis-h"
        />
      </li>
      <li
        className=""
      >
        <a
          href="category/test/page/100"
          onClick={[Function]}
        >
          100
        </a>
      </li>
    </ul>
  </div>
</section>
`;
